sdks/python/apache_beam/runners/portability/artifact_service.py

```diff
@@ -77,47 +87,120 @@ def PutArtifact(self, request_iterator, context=None):
       else:
         hasher.update(request.data.data)
         fout.write(request.data.data)
     fout.close()
     data_hash = hasher.hexdigest()
     if metadata.sha256 and metadata.sha256 != data_hash:
-      filesystems.FileSystems.delete([temp_path])
+      self._delete(temp_path)
       raise ValueError('Bad metadata hash: %s vs %s' % (
-          metadata.metadata.sha256, data_hash))
-    filesystems.FileSystems.rename(
-        [temp_path], [self._artifact_path(retrieval_token, metadata.name)])
+          metadata.sha256, data_hash))
+    self._rename(temp_path, artifact_path)
     return beam_artifact_api_pb2.PutArtifactResponse()
 
   def CommitManifest(self, request, context=None):
     retrieval_token = self.retrieval_token(request.staging_session_token)
-    with filesystems.FileSystems.create(
-        self._manifest_path(retrieval_token)) as fout:
-      fout.write(request.manifest.SerializeToString())
+    proxy_manifest = beam_artifact_api_pb2.ProxyManifest(
+        manifest=request.manifest,
+        location=[
+            beam_artifact_api_pb2.ProxyManifest.Location(
+                name=metadata.name,
+                uri=self._artifact_path(retrieval_token, metadata.name))
+            for metadata in request.manifest.artifact])
+    with self._open(self._manifest_path(retrieval_token), 'w') as fout:
+      fout.write(json_format.MessageToJson(proxy_manifest).encode('utf-8'))
     return beam_artifact_api_pb2.CommitManifestResponse(
         retrieval_token=retrieval_token)
 
   def GetManifest(self, request, context=None):
-    with filesystems.FileSystems.open(
-        self._manifest_path(request.retrieval_token)) as fin:
-      return beam_artifact_api_pb2.GetManifestResponse(
-          manifest=beam_artifact_api_pb2.Manifest.FromString(
-              fin.read()))
+    return beam_artifact_api_pb2.GetManifestResponse(
+        manifest=self._get_manifest_proxy(request.retrieval_token).manifest)
 
   def GetArtifact(self, request, context=None):
-    with filesystems.FileSystems.open(
-        self._artifact_path(request.retrieval_token, request.name)) as fin:
-      # This value is not emitted, but lets us yield a single empty
-      # chunk on an empty file.
-      chunk = True
-      while chunk:
-        chunk = fin.read(self._chunk_size)
-        yield beam_artifact_api_pb2.ArtifactChunk(data=chunk)
+    for artifact in self._get_manifest_proxy(request.retrieval_token).location:
+      if artifact.name == request.name:
+        with self._open(artifact.uri, 'r') as fin:
+          # This value is not emitted, but lets us yield a single empty
+          # chunk on an empty file.
+          chunk = True
+          while chunk:
+            chunk = fin.read(self._chunk_size)
+            yield beam_artifact_api_pb2.ArtifactChunk(data=chunk)
+        break
+    else:
+      raise ValueError('Unknown artifact: %s' % request.name)
+
+
+class ZipFileArtifactService(AbstractArtifactService):
```

Comment:
If this is the piece that depends on Python 3.6, could you document that here? And maybe add a check if you think it's needed?